casbin/distributed_enforcer.py

```diff
@@ -0,0 +1,134 @@
+from casbin import SyncedEnforcer
+import logging
+
+from casbin.persist import batch_adapter
+from casbin.model.policy_op import PolicyOp
+from casbin.persist.adapters import update_adapter
+
+
+class DistributedEnforcer(SyncedEnforcer):
+    """DistributedEnforcer wraps SyncedEnforcer for dispatcher."""
+
+    def __init__(self, model=None, adapter=None):
+        self.logger = logging.getLogger()
+        SyncedEnforcer.__init__(self, model, adapter)
+
+    def add_policy_self(self, should_persist, sec, ptype, rules):
+        """
+        AddPolicySelf provides a method for dispatcher to add authorization rules to the current policy.
+        The function returns the rules affected and error.
+        """
+        no_exists_policy = []
+        for rule in rules:
+            if self.get_model().has_policy(sec, ptype, rule):
+                no_exists_policy.append(rule)
+
+        if should_persist:
+            try:
+                if isinstance(self.adapter, batch_adapter):
+                    self.adapter.add_policies(sec, ptype, rules)
+            except Exception as e:
+                self.logger.log("An error occurred: " + e)
+
+        self.get_model().add_policies(sec, ptype, no_exists_policy)
+
+        if sec == "g":
+            try:
+                self.build_incremental_role_links(PolicyOp.Policy_add, ptype, no_exists_policy)
```

Comment:
@abichinger. For https://github.com/casbin/pycasbin/pull/130#discussion_r600690755 comment

Here `build_incremental_role_links` is defined in IntegerEnforcer class. The inheritance chain is : 

IntegerEnforcer -> ManagementEnforcer -> Enforcer -> SyncedEnforcer -> DistributedEnforcer. 

If SyncedEnforcer will not inherit Enforcer class, then DistributedEnforcer will not be able to use `build_incremental_role_links` method.

Reply:
> @abichinger. For [#130 (comment)](https://github.com/casbin/pycasbin/pull/130#discussion_r600690755) comment
> 
> Here `build_incremental_role_links` is defined in IntegerEnforcer class. The inheritance chain is :
> 
> IntegerEnforcer -> ManagementEnforcer -> Enforcer -> SyncedEnforcer -> DistributedEnforcer.
> 
> If SyncedEnforcer will not inherit Enforcer class, then DistributedEnforcer will not be able to use `build_incremental_role_links` method.

After creating this method is SyncedEnforcer there will be no need to inherit Enforcer in SyncedEnforcer so I'll remove this.